### PR TITLE
*: Add mimalloc feature to tikv_alloc

### PR DESCRIPTION
**Pull request by sticnarf**
Signed-off-by: Yilin Chen <sticnarf@gmail.com>

## What have you changed? (mandatory)

This PR adds [mimalloc](https://github.com/microsoft/mimalloc) support to TiKV. To use mimalloc as the allocator of TiKV, set `MIMALLOC` environment variable when building TiKV, like:

```bash
MIMALLOC=1 make dist_release
```

The mimalloc crate used is https://github.com/gnzlbg/mimallocator whose maintainer is the same as jemallocator's.

## What are the type of the changes? (mandatory)

- New feature (change which adds functionality

## How has this PR been tested? (mandatory)

Untested

## Does this PR affect documentation (docs) or release note? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

## Benchmark result if necessary (optional)

A little improvement under sysbench while a little regression under go-ycsb. It should be investigated more.

image
image

### CODE REVIEW

**Comment by siddontang** [comment]
Great work!!!

Do we statically link mimalloc?

**Comment by siddontang** [comment]
does mimalloc support outputting statistics like jemalloc?

**Comment by kennytm** [comment]
You could enable the `stats` feature of the mimalloc dependency to print the stats on exit.

File: components/tikv_alloc/Cargo.toml [hunk 1]
```diff
@@ -44,3 +45,7 @@ optional = true
 version = "0.3.0"
 optional = true
 features = ["bundled"]
+
+[dependencies.mimallocator]
+version = "0.1.3"
+optional = true
```
Comment (breeswish) [inline components/tikv_alloc/Cargo.toml]: trailing new line

**Comment by sticnarf** [comment]
> Do we statically link mimalloc?

Yes. I believe it is statically linked.

**Comment by sticnarf** [comment]
Now, `stats` feature is enabled.

**Comment by siddontang** [comment]
do we need to implement fetch_stats and dump_stats？

**Comment by sticnarf** [comment]
@siddontang The `stats` feature just prints statistics to `stderr` when the program exits.
The `mimallocator` crate does not support runtime statistics yet but it is possible to call `mi_stats_print` to collect statistics.

**Comment by sticnarf** [comment]
After a second test,  I am confident it's not a mistake that sysbench and go-ycsb give different results. 

I think it just tells us mimalloc does not work better in every case. When value size differs, allocation size and frequency can differ too. No allocator can handle every case well.

**Comment by kennytm** [comment]
(Just to clarify I don't mean the `stats` feature should always be enabled 😅)

**Comment by breeswish** [comment]
Maybe related: according to https://github.com/microsoft/mimalloc/issues/11 mimallocator was not built for many short lived small allocations.

**Comment by sticnarf** [comment]
> (Just to clarify I don't mean the `stats` feature should always be enabled )

Sorry, at first I thought the stats feature of mimalloc is similar with that of jemalloc :sweat_smile:

**Comment by siddontang** [comment]
CI failed @sticnarf

**Pull request merged**